feat: implement hotline pool system and optimize voice agent performance

arx-optimus-17 opened the pull request:
- Added database-backed hotline pool for MIA voice agents, reducing response time to ~31ms
- Updated API endpoint to use pool-first logic with graceful fallback to on-demand creation
- Introduced ENABLE_HOTLINE config for feature toggle and instant rollback
- Refactored service, database, and migration scripts for pool lifecycle managemefeat: implement hotline pool system and optimize voice agent performance
- Developer Proof: https://drive.google.com/file/d/11cDXIuH05E6W8iadsM7cnpUMlZyl1e_r/view?usp=sharing

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- New Features
  - Faster voice agent connections via a pre-allocated Daily Hotline pool (with graceful fallback).
  - New endpoints: GET /hotline/status and POST /hotline/release/{room_id}.
  - New WebSocket endpoint: /ws/live for Gemini Live.
- Configuration
  - New environment toggles to enable and tune the hotline pool; startup validation and informative logging.
- Performance & Reliability
  - Background pool management, expired-room cleanup, and optional agent cleanup on shutdown.
- Database
  - Added hotline rooms table creation; included in the “create all” flow.
- Chores
  - Added psutil dependency for process management.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

badri-singhal commented on app/ws/live_session.py:
```diff
@@ -439,6 +439,20 @@ async def forward_from_gemini():
         await close_gemini_session(gemini_session_cm)
         
+        # Release hotline room if session_id is available
+        if hasattr(websocket.state, 'hotline_session_id'):
```
@arx-optimus-17 bhaiya, revert these changes, this is not actaully being used in the production. In some time we will remove these files.

badri-singhal commented on app/main.py:
```diff
@@ -230,11 +266,92 @@ async def telephony_websocket_handler(serviceIdentifier: str, workflow: str, web
+        raise HTTPException(status_code=500, detail=f"Failed to associate session: {str(e)}")
+
+
+@app.get("/health")
```
we already have this endpoint

badri-singhal commented on app/main.py:
```diff
@@ -376,6 +588,115 @@ async def get_version():
+        raise HTTPException(status_code=404, detail="Hotline system is disabled")
+    
+    try:
+        from uuid import UUID
```
move this to top

badri-singhal commented on app/main.py:
```diff
@@ -376,6 +588,115 @@ async def get_version():
+        room_uuid = UUID(room_id)
+        
+        async for conn in get_db_connection():
+            from app.database.queries.hotline import release_room_query
```
move this to top

badri-singhal commented on app/main.py:
```diff
@@ -376,6 +588,115 @@ async def get_version():
+        })
+
+# Room release endpoint for proper lifecycle management
+@app.post("/hotline/release/{room_id}")
```
will lighthouse call this endpoint?

badri-singhal replied:
Also, should'nt this be authenticated?

badri-singhal commented on app/main.py:
```diff
@@ -376,6 +588,115 @@ async def get_version():
     return JSONResponse({"version": __version__})
 
+# Hotline debugging endpoint
+@app.get("/hotline/debug")
```
This endpoint is not required

badri-singhal commented on app/main.py:
```diff
@@ -230,11 +266,92 @@ async def telephony_websocket_handler(serviceIdentifier: str, workflow: str, web
+    """
+    try:
+        # Find the WebSocket connection with matching token in query params
+        from app.ws.live_session import get_active_connections
```
move this to top

badri-singhal commented on app/main.py:
```diff
@@ -230,11 +266,92 @@ async def telephony_websocket_handler(serviceIdentifier: str, workflow: str, web
+        
+        # Get current pool stats
+        async for conn in get_db_connection():
+            from app.database.queries.hotline import get_pool_stats_query
```
move this to top of the file

badri-singhal commented on app/main.py:
```diff
@@ -376,6 +588,115 @@ async def get_version():
+        })
+
+# Hotline status endpoint
+@app.get("/hotline/status")
```
`/hotline/status` This endpoint is already defined above, why two times

arx-optimus-17 committed:
feat: implement hotline pool system and optimize voice agent performance

- Added database-backed hotline pool for MIA voice agents, reducing response time to ~31ms
- Updated API endpoint to use pool-first logic with graceful fallback to on-demand creation
- Introduced ENABLE_HOTLINE config for feature toggle and instant rollback
- Refactored service, database, and migration scripts for pool lifecycle managemefeat: implement hotline pool system and optimize voice agent performancenagemefeat: implement hotline pool system and optimize voice agent performance

badri-singhal commented on app/main.py:
```diff
@@ -230,11 +270,87 @@ async def telephony_websocket_handler(serviceIdentifier: str, workflow: str, web
     await handle_websocket_session(websocket)
 
+
+@app.post("/hotline/associate-session")
```
.

swaroopvarma1 commented on app/schemas.py:
```diff
@@ -22,6 +22,17 @@ class RequestedBy(str, Enum):
 class Workflow(str, Enum):
     ORDER_CONFIRMATION = "order-confirmation"
 
+class HotlineRoomStatus(str, Enum):
```
DailyRoomStatus

swaroopvarma1 commented on app/database/queries/hotline.py:
```diff
@@ -0,0 +1,195 @@
+"""
```
daily-hotline.py

swaroopvarma1 commented on app/core/config.py:
```diff
@@ -115,6 +115,29 @@ def get_required_env(var_name: str) -> str:
 KEEP_RECENT_TURNS = int(os.environ.get("KEEP_RECENT_TURNS", 2))
 
+# Hotline Configuration - Room Reservation System
+ENABLE_HOTLINE = os.environ.get("ENABLE_HOTLINE", "false").lower() == "true"
```
ENABLE_HOTLINE -> ENABLE_daily_HOTLINE

swaroopvarma1 commented on app/core/config.py:
```diff
@@ -115,6 +115,29 @@ def get_required_env(var_name: str) -> str:
+HOTLINE_POOL_MIN_SIZE = int(os.environ.get("HOTLINE_POOL_MIN_SIZE", "3"))
+HOTLINE_POOL_MAX_SIZE = int(os.environ.get("HOTLINE_POOL_MAX_SIZE", "8"))
+HOTLINE_ROOM_EXPIRY_MINUTES = int(os.environ.get("HOTLINE_ROOM_EXPIRY_MINUTES", "30"))
+HOTLINE_ENABLE_ON_DEMAND = os.environ.get("HOTLINE_ENABLE_ON_DEMAND", "true").lower() == "true"
```
this should be always true - no need for an env

swaroopvarma1 commented on app/core/config.py:
```diff
@@ -115,6 +115,29 @@ def get_required_env(var_name: str) -> str:
+HOTLINE_CLEANUP_ON_SHUTDOWN = os.environ.get("HOTLINE_CLEANUP_ON_SHUTDOWN", "false").lower() == "true"
+HOTLINE_POOL_MIN_SIZE = int(os.environ.get("HOTLINE_POOL_MIN_SIZE", "3"))
+HOTLINE_POOL_MAX_SIZE = int(os.environ.get("HOTLINE_POOL_MAX_SIZE", "8"))
+HOTLINE_ROOM_EXPIRY_MINUTES = int(os.environ.get("HOTLINE_ROOM_EXPIRY_MINUTES", "30"))
```
Default -> 45 mins

swaroopvarma1 commented on app/core/config.py:
```diff
@@ -115,6 +115,29 @@ def get_required_env(var_name: str) -> str:
+# Hotline Configuration - Room Reservation System
+ENABLE_HOTLINE = os.environ.get("ENABLE_HOTLINE", "false").lower() == "true"
+HOTLINE_CLEANUP_ON_SHUTDOWN = os.environ.get("HOTLINE_CLEANUP_ON_SHUTDOWN", "false").lower() == "true"
+HOTLINE_POOL_MIN_SIZE = int(os.environ.get("HOTLINE_POOL_MIN_SIZE", "3"))
```
Add daily prefix for all env

swaroopvarma1 commented on app/core/config.py:
```diff
@@ -115,6 +115,29 @@ def get_required_env(var_name: str) -> str:
+        raise ValueError("HOTLINE_POOL_MAX_SIZE must be greater than or equal to HOTLINE_POOL_MIN_SIZE")
+    if HOTLINE_ROOM_EXPIRY_MINUTES <= 0:
+        logger.error("HOTLINE_ROOM_EXPIRY_MINUTES must be greater than 0")
+        raise ValueError("HOTLINE_ROOM_EXPIRY_MINUTES must be greater than 0")
```
for these cases fallback to default values instead of error

swaroopvarma1 commented on app/database/queries/hotline.py:
```diff
@@ -0,0 +1,195 @@
+
+
+# Table names
+HOTLINE_ROOMS_TABLE = "hotline_rooms"
```
daily_hotline_rooms

swaroopvarma1 commented on app/scripts/create_tables.py:
```diff
@@ -5,11 +5,13 @@
 
 # Table names
 CALL_DATA_TABLE = "call_data"
+HOTLINE_ROOMS_TABLE = "hotline_rooms"
```
daily_hotline_rooms

swaroopvarma1 commented on app/services/hotline_manager.py:
```diff
@@ -0,0 +1,624 @@
+        # Check voice compatibility - if user requests non-default voice, force on-demand creation
+        requested_voice = request_params.get("voice_name")
+        requested_tts = request_params.get("tts_provider", "google")
+        default_voice_name = "MIA"  # Our pool default voice name (enum value)
```
make hotline available for all voices

swaroopvarma1 commented on app/services/hotline_manager.py:
```diff
@@ -0,0 +1,624 @@
+        """Spawn agent subprocess in pool mode (waiting for connection)."""
+        try:
+            # Build command args similar to existing bot_connect
+            bot_file = "app.agents.voice.automatic"
```
subprocess init should always happen from main

swaroopvarma1 commented on app/services/hotline_manager.py:
```diff
@@ -0,0 +1,624 @@
+"""
```
services/automatic/daily/hotline_manager.py

swaroopvarma1 commented on app/core/config.py:
```diff
@@ -115,6 +115,29 @@ def get_required_env(var_name: str) -> str:
 
+# Daily Hotline Configuration - Room Reservation System
+ENABLE_DAILY_HOTLINE = os.environ.get("ENABLE_DAILY_HOTLINE", "false").lower() == "true"
+DAILY_HOTLINE_CLEANUP_ON_SHUTDOWN = os.environ.get("DAILY_HOTLINE_CLEANUP_ON_SHUTDOWN", "false").lower() == "true"
```
This should be default true

swaroopvarma1 commented on app/core/config.py:
```diff
@@ -115,6 +115,29 @@ def get_required_env(var_name: str) -> str:
+DAILY_HOTLINE_POOL_MIN_SIZE = int(os.environ.get("DAILY_HOTLINE_POOL_MIN_SIZE", "3"))
+DAILY_HOTLINE_POOL_MAX_SIZE = int(os.environ.get("DAILY_HOTLINE_POOL_MAX_SIZE", "8"))
+DAILY_HOTLINE_ROOM_EXPIRY_MINUTES = int(os.environ.get("DAILY_HOTLINE_ROOM_EXPIRY_MINUTES", "45"))
+DAILY_HOTLINE_ENABLE_ON_DEMAND = True  # Always enabled
```
We can remove DAILY_HOTLINE_ENABLE_ON_DEMAND this and directly write logic assuming it is always true

swaroopvarma1 commented on app/database/queries/daily_hotline.py:
```diff
@@ -0,0 +1,222 @@
+    if session_id:
+        text = f"""
+            INSERT INTO "{HOTLINE_ROOMS_TABLE}" 
+            ("daily_room_url", "daily_token", "status", "agent_pid", "expires_at", "session_id")
```
agent_pid might not be needed to track

swaroopvarma1 commented on app/schemas.py:
```diff
@@ -22,6 +22,17 @@ class RequestedBy(str, Enum):
+    """Status enum for hotline room reservation system."""
+    AVAILABLE = "available"
+    RESERVED = "reserved" 
+    IN_USE = "in_use"
```
Expired not needed?

swaroopvarma1 commented on app/services/automatic/daily/hotline_manager.py:
```diff
@@ -0,0 +1,596 @@
+                        await asyncio.sleep(sleep_time)
+                        break
+                except Exception as e:
+                    logger.error(f"❌ Error in pool maintenance cycle: {e}")
```
no emojis pls

swaroopvarma1 commented on app/services/automatic/daily/hotline_manager.py:
```diff
@@ -0,0 +1,596 @@
+                self._startup_cleanup_done = True
+                logger.info("✅ Startup cleanup completed")
+            
+            while True:
```
Instead of infinite loop, can we do like below?
- Whenever a request is received let's check and scale
- When Pipeline has to be cancelled, which we can track using on_pipeline_cancelled, can we reset room to available?
- Again when a request comes, can we scale this accordingly?

badri-singhal commented:
Rebase with release, and run ./scripts/setup.sh file before committing
